### ISSUE TITLE
Wrong behaviour of set/update on small ecirca

### ISSUE THREAD

**Issue by si14** [band115/ecirca#31]
Strange atoms are returned as an old value in set/update of small ecirca.

**Comment by si14** [comment]
Temporary workaround: small ecirca = medium ecirca for now.

**Comment by si14** [comment]
It's not helping at all (too much changes to get things work right with this fix). Here is some informations for those who want to help:

```
[si14@si14-laptop temp]$ git clone git://github.com/band115/ecirca.git && cd ecirca
Cloning into ecirca...
...
[si14@si14-laptop ecirca]$ rebar get-deps compile
==> ecirca (get-deps)
...
==> ecirca (compile)
...
Compiling c_src/ecirca_large.c
Compiling c_src/ecirca_medium.c
Compiling c_src/ecirca_small.c
[si14@si14-laptop ecirca]$ ./start-dev.sh 
Erlang R14B03 (erts-5.8.4) [source] [64-bit halfword] [smp:4:4] [rq:4] [async-threads:10] [kernel-poll:true]
...
=PROGRESS REPORT==== 14-Sep-2011::21:46:37 ===
         application: sasl
          started_at: circa_dev@localhost
Eshell V5.8.4  (abort with ^G)

(circa_dev@localhost)1> {ok, T1} = ecirca:new(5, last, small).
{ok,{ecirca,#Ref<0.0.0.57>,<<>>,<0.48.0>,small}}
(circa_dev@localhost)2> {ok, T2} = ecirca:new(5, last, medium).
{ok,{ecirca,#Ref<0.0.0.63>,<<>>,<0.48.0>,medium}}
(circa_dev@localhost)3> ecirca:slice(T1, 1, 5).
{ok,[empty,empty,empty,empty,empty]}
(circa_dev@localhost)4> ecirca:slice(T2, 1, 5).
{ok,[empty,empty,empty,empty,empty]}
(circa_dev@localhost)5> ecirca:set(T1, 1, 2).
{ok,{signal,2}}  <-- this 'signal' should be 'empty'!
(circa_dev@localhost)6> ecirca:set(T2, 1, 2).
{ok,{empty,2}}
(circa_dev@localhost)7> ecirca:slice(T1, 1, 5).
{ok,[2,empty,empty,empty,empty]}
(circa_dev@localhost)8> ecirca:slice(T2, 1, 5).
{ok,[2,empty,empty,empty,empty]}
(circa_dev@localhost)9> ecirca:set(T1, 1, empty).
{ok,{2,signal}} <-- here again
```

As you can see, ecirca:slice shows right atoms, ecirca:set shows very strange behaviour. As one can see uncommenting printfs here https://github.com/band115/ecirca/blob/8eea958f25fe3175ac259053ed2903c6ebd1af62/priv/ecirca.c#L562 , all arguments and returned atom (being casted to number) are identical in the case of slice and set, so one can suppose there should be identical atoms, but they are different.

**Comment by si14** [comment]
There is a report that this issue won't reproduce without halfword emulator. Checking now.

**Comment by si14** [comment]
Recompiled Erlang without halfword, bug is still here:

```
Erlang R14B03 (erts-5.8.4) [source] [64-bit] [smp:4:4] [rq:4] [async-threads:10] [hipe] [kernel-poll:true]

(circa_dev@localhost)1> {ok, T} = ecirca:new(5, last, small).
{ok,{ecirca,#Ref<0.0.0.69>,<<>>,<0.52.0>,small}}
(circa_dev@localhost)2> ecirca:set(T, 1, 2).                 
{ok,{sinh,2}}
(circa_dev@localhost)3> ecirca:slice(T, 1, 5).
{ok,[2,empty,empty,empty,empty]}
(circa_dev@localhost)4> ecirca:set(T, 1, empty).
{ok,{2,sinh}}
```

**Comment by si14** [comment]
Looks like bug is not present on x32 machines: http://paste.org/pastebin/view/38420

**Comment by si14** [comment]
R14B02 is affected:

```
Erlang R14B02 (erts-5.8.3) [source] [64-bit] [smp:4:4] [rq:4] [async-threads:10] [hipe] [kernel-poll:true]
...
Eshell V5.8.3  (abort with ^G)
(circa_dev@localhost)1> {ok, T} = ecirca:new(5, last, small).
{ok,{ecirca,#Ref<0.0.0.56>,<<>>,<0.48.0>,small}}
(circa_dev@localhost)2> ecirca:set(T, 1, 2).
{ok,{resume_process,2}}
(circa_dev@localhost)3> ecirca:set(T, 1, empty).
{ok,{2,resume_process}}
```

**Comment by si14** [comment]
fixed in d681df1
